### ISSUE TITLE
Adding equivalences during rewriting

### ISSUE THREAD

**Issue by bjchambers** [egraphs-good/egg#126]
I'm not exactly sure what terms to use to describe this, so I'll explain the example and some thoughts.

Say we have a simple language that includes some math, and the ability to construct records `{ (<name>: <expr>)* }` and access fields `<expr>.<name>`. A very reasonable rewrite rule in this would be `{ ?name: ?expr, .. }.?name => ?expr`. Obviously, we can't implement that rule for a variety of reasons, including the fact it is variadic.

Now, one option would be to match the field ref, and have that run a specially implemented rule that used information about the expression (possibly whether it was a record) to apply the rewrite. I'll probably pursue if necessary. But, that also has some problems -- for instance, if there are other operators such as an `if` expression then the field ref may not be applied *directly*.

Another idea I had was to focus on the record introduction rather than the field ref. If for every record `{ ?name1: ?expr1, ?name2: ?expr2, ... }` I added `?record.?name1 = ?expr1`, `?record.?name2 = ?expr2`, etc. then I believe that those equivalences would be preferred, etc.

But, I'm not sure how to go about writing a rule that says "for each record (ideally only once when it is added), add these equivalences". I think it is possible to explicitly union things within the DFG, so adding the equivalences is easy. It seems like it may be possible with a custom searcher and a custom applier, but I'm not sure if that would re-execute the search every time.

I'm also not sure if it should produce a single subst for the record (and then the applier should iterate over the fields) or if it should produce a subst for each field. The latter seems tricky since there is no way to carry the field name in the subst.

Are there any examples of applying egg to languages with records with these kinds of equivalences?

**Comment by bjchambers** [comment]
> > for instance, if there are other operators such as an if expression then the field ref may not be applied directly.

> I did not understand that comment, is your language referentially opaque?

That was worded poorly. I meant something like `if (...) { { a: an_expr, ... }.a } else { { b: an_expr}.b }`. The result of the `if` expression is the same because both branches ar equivalent to `an_expr`.

> Maybe an e-class analysis with modify action would be suitable here? This looks a bit like constant folding (record projection partial evaluation).

I think that could work. When is the the `modify` action called? All the time, or only after a node is created/merged?

**Comment by bjchambers** [comment]
Yep. That works. Thanks!

It feels a bit like it may be doing wasted work -- every time modify is called it has to iterate over the nodes (whether or not they are new) and create field-ref nodes for each field in any records found.

I wonder if there is any way to (1) only do it for nodes that are newly part of the class and/or (2) only do it for field-refs that exist in the graph.

I tried to do (2) using lookup but that fails -- I'm guessing because a field ref may be added without modifying the record, so it may come into existance after the equivalences are added. It seems like adding all field refs isn't too much extra work, but (1) has me a bit worried.

**Comment by bjchambers** [comment]
Yeah, I think that is an alternative. Using `modify` in the analysis is pretty straightforward since it is basically using the egraph to record ".foo" is equivalent to this expression. If I did it in analysis, it requires a lot more plumbing -- every node needs to be annotated with whether it "has fields", and the e-class of those "fields". Once that plumbing was done, you could write the rule that is basically "field ref of a thing that the analysis says has fields is equivalent to the things the analysis says that field is equivalent to".

**Comment by bjchambers** [comment]
Commenting to add some more details. I *think* that a hybrid approach may be necessary. The problem is that *just* adding equivalences for fields of a record expression isn't enough if you have any functions that "pass the record through" (eg., that return a record based on an input record). Ideally, you'd like to be able to record equivalences for the result of such functions.

I then considered *just* adding the map as suggested, but I think that *also* suffers. Ideally, you'd want to be able to say "if two expressions are equivalent, then the IDs associated with their fields of the same name are also equivalent". But the merge operation can't register *additional* equivalences.

So I think what I'm trying next will be for analysis to populate and maintain the map of field names to equivalent expressions, but then having the `modify` register equivalences for each of those fields. I believe that what will happen is if I have `foo.x = 5` and `bar.x = y`, then when I merge `foo` and `bar` it will also enable merging `foo.x` and `bar.x`, which will then merge `5` and `y`.

Curious if there are other examples/ideas for handling records/field accesses. My goal is to be able to remove the record creation/field access when possible.

**Comment by bjchambers** [comment]
One thing I'm curious about -- it seems inefficient to (re)create these equivalences each time `modify` is called. Is there any way to determine which nodes are "new" to the `eclass`? That would allow me to visit only the new nodes.

**Comment by bjchambers** [comment]
One way I *could* accomplish this is having a list of "new" nodes. When `b` is merged into `a` I could add `b` (and everything in `b`s work list) to `a`. Then in `modify` I could look at the list, and process anything that is in there (and then empty it). Feels dirty, but could work...

**Comment by mwillsey** [comment]
See #134, I think adding `&mut self` to `Analysis::merge` might work.

**Comment by bjchambers** [comment]
I think that would work. So the idea would be that the analysis data could contain "this node has fields equivalent to these IDs". When two nodes with that data are merged, it could union each of the equivalent fields. This would be possible with the `&mut self` in the `Analysis::merge`.

Similarly, I could have a rewrite rule that handles field ref -- if the `record` has field information from analysis, then it could rewrite `record.field` as equivalent to whichever field the analysis had. I think this would require a custom searcher (since there isn't a single field access node) and a custom applier, but that should work. I'll keep an eye on #134 and maybe give it a shot when that goes in.

**Comment by mwillsey** [comment]
Sounds good! FYI #134 is now resolved so you should be good to go :)
